### PR TITLE
feat(pageserver): generate basebackup from aux file v2 storage

### PR DESCRIPTION
**Pull request by skyzh**
## Problem

This pull request adds the new basebackup read path + aux file write path. In the regression test, all logical replication tests are run with matrix aux_file_v2=false/true.

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [x] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

**Comment by skyzh** [comment]
All failed tests pass locally so I assume it's something wrong with vectored get env variable in CI. Fix in https://github.com/neondatabase/neon/pull/7569.

**Comment by skyzh** [comment]
rebased to the new scan implementation pull request and added aux_file_v2 test matrix for regress tests.

File: pageserver/src/pgdatadir_mapping.rs [hunk 1]
```diff
@@ -675,15 +675,31 @@ impl Timeline {
...
-                Ok(HashMap::new())
+        if self.get_switch_to_aux_file_v2() {
+            let kv = self
+                .scan(KeySpace::single(Key::metadata_aux_key_range()), lsn, ctx)
+                .await
+                .context("scan")?;
+            let mut result = HashMap::new();
+            for (_, v) in kv {
+                let v = v.context("get value")?;
+                let v = aux_file::decode_file_value(&v).context("value decode")?;
```
Comment (VladLazar) [inline pageserver/src/pgdatadir_mapping.rs:686]: Do you intend to keep these contexts or are they for debugging only?

**Review by VladLazar — COMMENTED** [review]
The looks good to me, but I have questions about the testing.
Assuming a bug with aux files v2, I don't think there is a guarantee that postgress will generate an error. Have you considered validating the two aux files implementations against each other (at least in the tests)?

File: pageserver/src/pgdatadir_mapping.rs (hunk 1 above)
Comment (skyzh) [reply pageserver/src/pgdatadir_mapping.rs:686]: I want to keep them

**Comment by skyzh** [comment]
Added a new cross-validation mode to aux file policy flag, waiting for CI runs.

**Comment by skyzh** [comment]
added https://github.com/neondatabase/neon/pull/7517/commits/26efd47e749632c1b657e6048f26536e89a82a24 to fix a bug on the code path of using vectored get for sequential get.

Think about the case if a key does not exist. In the code path, we will create a ValueReconstructState of empty image and empty WAL, which then gets passed to `reconstruct_value`, and throwing `"base image for {key} at {request_lsn} not found"` PageReconstructError::Other error. This case is actually a missing key error.

Therefore, if the cached image is not found, we do not add it to the value reconstruct state to avoid calling into reconstruct_value.

It's unknown why this case only happens when v1+v2 are both enabled. Probably due to the difference of layer map in two cases, where in one case, the vectored get code path could catch the missing key error earlier than the point that reconstruct_value gets called.

**Review by VladLazar — APPROVED** [review]
Looking good. Just a few nits.

File: pageserver/src/pgdatadir_mapping.rs [hunk 2]
```diff
@@ -1404,89 +1462,121 @@ impl<'a> DatadirModification<'a> {
...
+                    }
                 }
             }
-        }
 
-        self.pending_directory_entries
-            .push((DirectoryKind::AuxFiles, n_files));
+            self.pending_directory_entries
+                .push((DirectoryKind::AuxFiles, n_files));
+        }
```
[the remark this answers is not in the document]
Comment (skyzh) [reply pageserver/src/pgdatadir_mapping.rs:1579]: no, the init state of aux file v2 is simply an empty keyspace. `init_aux_dir` will be removed once we decide to remove the v1 code path.

File: pageserver/src/pgdatadir_mapping.rs [hunk 3]
```diff
@@ -688,6 +689,63 @@ impl Timeline {
...
+                    }
+                    (Ok(_), Err(v2)) => {
+                        tracing::error!("aux file v1 returns Ok while aux file v2 returns an err");
+                        Err(v2)
+                    }
+                    (Err(v1), Ok(_)) => {
+                        tracing::error!("aux file v2 returns Ok while aux file v1 returns an err");
+                        Err(v1)
+                    }
+                    (Err(v1), Err(_)) => Err(v1),
```
[the remark this answers is not in the document]
Comment (skyzh) [reply pageserver/src/pgdatadir_mapping.rs:743]: Changed to return Err(v2).

In the case of error parity, I don't see test cases testing on errors of list_aux_files. Furthermore, as they operate on different kinds of storage format, the error will likely be different. For example, missing key for v1 would be the AUX_FILE_KEYS, while missing key for v2 would be the encoded key. The same applies to the redo error, where aux file v2 does not do any redos.

File: pageserver/src/pgdatadir_mapping.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (VladLazar) [reply pageserver/src/pgdatadir_mapping.rs:1579]: I was alluding to the fact that you can skip the calls if using aux files v2.

[the remark this answers is not in the document]
Comment (skyzh) [reply pageserver/src/pgdatadir_mapping.rs:1582]: resolved in https://github.com/neondatabase/neon/pull/7517/commits/801550e33687cc0732df7f9a889688c89f004c26

**Pull request merged**